Add Examples and usecase for Lag windows function (#5529)
Codecademy/docs: content/sql/concepts/window-functions/terms/lag/lag.md

- `partition_by` - Allows the result set to be grouped based on a column. This is an optional argument, if not specified the result set will be treated as a single group.
- `order_by` - Determines the order of the result set. If `partition_by` is specified, it will order the grouped data instead.

## Examples

> The examples below use the table Sales, which contains orders of different products that are sold on a particular date and their price that they are sold to.

Reviewer: We should use the table name inside backticks to signify the exact name used in the code as well. Also, we should use the word table after the table name for better readability.

Also, some minor adjustments can make it much easier to follow like this -

```suggestion
The examples below use the Sales table, which contains orders for different products sold on specific dates, along with their prices.
```

Sales Table

| product_id | order_date | sold_price |
| ---------- | ---------- | -------- |
|          1 | 2024-01-01 |      100 |
|          1 | 2024-02-01 |      120 |
|          1 | 2024-03-01 |       80 |
|          2 | 2024-01-01 |       50 |
|          2 | 2024-02-01 |       60 |
|          2 | 2024-02-02 |       60 |

Example 1 : 

> The below example uses `LAG()` on column `order_date` with Order By `order_date`.

Reviewer: We can simplify the flow of the sentence by replacing `Order by` with ordered by, as follows -

```suggestion
The below example uses `LAG()` on column `order_date`, ordered by `order_date`.
```

```sql
SELECT
    product_id,
    order_date,
    sold_price,
    LAG(order_date) OVER (
      ORDER BY order_date
      ) AS previous_order_date
FROM
    sales;
```

> The output is a table with a new column `previous_order_date`, which contains previous entries of `order_date` in the table when ordered by `order_date`. The offset is set to 1 by default, and NULL is displayed when the offset is out of range.

Reviewer: It's better to clarify the nature of the `LAG()` function's behavior when the offset is out of range, i.e., when the previous_order_date is NULL.

```suggestion
The output is a table with a new column `previous_order_date`, which contains previous entries of `order_date` in the table when ordered by `order_date`. The default offset is set to 1, meaning the previous row’s order_date will be displayed. When there is no previous row (such as for the first entry in the table), NULL will be displayed.
```

Output

| product_id | order_date | sold_price | previous_order_date |
| ---------- | ---------- | ---------- | ------------------- |
|          1 | 2024-01-01 |        100 | NULL                |
|          2 | 2024-01-01 |         50 | 2024-01-01          |
|          1 | 2024-02-01 |        120 | 2024-01-01          |
|          2 | 2024-02-01 |         60 | 2024-02-01          |
|          2 | 2024-02-02 |         60 | 2024-02-01          |
|          1 | 2024-03-01 |         80 | 2024-02-02          |

Example 2 :

> The below example uses `LAG()` with offset of 1, partitioned by `product_id` and Order by `order_date`.

Reviewer: We can simplify the flow of the sentence by replacing Order by with ordered by, as follows -

```suggestion
The below example uses `LAG()` with offset of 1, partitioned by `product_id` and ordered by `order_date`.
```

```sql
SELECT
    product_id,
    order_date,
    sold_price,
    LAG(sold_price,1) OVER (
      PARTITION BY product_id 
      ORDER BY order_date
      ) AS previously_sold_price
FROM
    sales;
```

> The output is a table that features a new column `previously_sold_price`, which holds the sold price of the product on the previous sold date. The first record of every `product_id` is NULL because a default was not specified as the previous date is out of range for the first sale of the product.

Reviewer: `sold` is the wrong word to be used in the sentence here, rather we should call it sale price or else we should use the exact column name inside backticks for reference to the table like this - `sold_price`.

The second sentence here is a bit difficult to follow, simplify it as follows -

```suggestion
The output is a table that features a new column `previously_sold_price`, which holds the sale price of the product on the previous sale date. For the first record of each product_id, NULL is displayed because there is no previous sale for that product.
```

Output

| product_id | order_date | sold_price | previously_sold_price |
| ---------- | ---------- | ---------- | --------------------- |
|          1 | 2024-01-01 |        100 |                  NULL |
|          1 | 2024-02-01 |        120 |                   100 |
|          1 | 2024-03-01 |         80 |                   120 |
|          2 | 2024-01-01 |         50 |                  NULL |
|          2 | 2024-02-01 |         60 |                    50 |
|          2 | 2024-02-02 |         60 |                    60 |

Example 3:

The below example uses `LAG()` with offset of 1 and with default value "First Sale".

```sql
SELECT
    product_id,
    order_date,
    sold_price,

>     LAG(product_id,1,"First Sale") OVER (

Reviewer: In SQL, single quotes are used to denote string constants, i.e., values that represent text. For example: 'First Sale' is a string constant containing the text First Sale.

```suggestion
    LAG(product_id, 1, 'First Sale') OVER (
```

      ORDER BY order_date
      ) AS previous_sold_product
FROM
    sales;
  
```

> This outputs column `previous_sold_product`, which contains the previous sold product for each sale ordered by `order_date`. As there is no previous sale for the first entry, the default is set to "First Sale".

Reviewer: We can simplify the explanation by specifying that the Product ID is returned in the `previous_sold_product` column and adding more detail to explain when the default value is applied, as follows -

```suggestion
This query outputs a new column, `previous_sold_product`, which contains the product ID from the previous sale for each row ordered by order_date. If there is no previous sale (for the first entry), the default value "First Sale" is used.
```
